src/Microsoft.VisualStudio.Threading/AsyncReaderWriterLock.cs

```diff
@@ -2272,6 +2286,11 @@ public void OnCompleted(Action continuation)
 
                 this.cancellationRegistration = this.cancellationToken.Register(CancellationResponseAction, this, useSynchronizationContext: false);
                 this.lck.PendAwaiter(this);
+
+                if (this.cancellationToken.IsCancellationRequested && this.cancellationRegistration == default(CancellationTokenRegistration))
```

Comment:
If I read this condition right, it means "if the token was canceled *after* we called Register above such that the delegate will be (or may have been) invoked asynchronously..."
Is that right?
If so, it's not immediately clear to me why we need to invoke the cancellation responder ourselves here if it will be or has been invoked already.

Reply:
The race condition happens that the CancellationToken is triggered before line 2287.  In that case, cancellationToken.Register will not register any callback, but will it within that function call.  Our CancellationToken handler only handles Cancellation, if the request is in the queue.  However, the request is only added to the queue at line 2288, so it becomes no-op.  The new code only kicks in this condition, (cancelled, but cancellation register is empty, and triggers the cancellation logic again.  This will fix the problem.  If the cancellation token is triggered after 2287, the cancellationRegister will contain something (it won't be cleaned up even after it is disposed), so we will not call into this new logic under that condition.

In the earlier version, I moved the cancellation registration after we add the request to the queue.  That will fix this dead lock issue, but there is a chance the cancellationRegistration can be leaked.  (If the lock is issued immediately, it may call into OnCompleted call back on a different thread, so the registration can happen after the lock has been issued to the consumer code, and will no longer be cleaned up.